client/tracing/src/logging/layers/prefix_layer.rs

```diff
@@ -0,0 +1,86 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use tracing::{span::Attributes, Id, Subscriber};
+use tracing_subscriber::{layer::Context, registry::LookupSpan, Layer};
+
+/// Span name used for the logging prefix. See macro `sc_tracing::logging::prefix_logs_with!`
+pub const PREFIX_LOG_SPAN: &str = "substrate-log-prefix";
+
+/// A `Layer` that captures the prefix span ([`PREFIX_LOG_SPAN`]) which is then used by
+/// [`EventFormat`] to prefix the log lines by customizable string.
+///
+/// See the macro `sc_cli::prefix_logs_with!` for more details.
+pub struct PrefixLayer;
+
+impl<S> Layer<S> for PrefixLayer
+where
+	S: Subscriber + for<'a> LookupSpan<'a>,
+{
+	fn new_span(&self, attrs: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
+		let span = ctx
+			.span(id)
+			.expect("new_span has been called for this span; qed");
```

Comment:
`new_span` is called when a new span is registered. So it should be there.

But maybe I'm missing something. I don't know tracing very well. The context could be uninitialized you say?

Reply:
I'd have to dig in deeper to be sure how `new_span` is called to tell if `Context` can ever have `subscriber` set to `None` in our usage. [Docs link](https://docs.rs/tracing-subscriber/0.2.15/tracing_subscriber/layer/struct.Context.html#method.span). The proof string seems sketchy to me though.

Reply:
@cecton if you can not proof this, please do not use such a proof in the code. A proof means, you  are 100% that this can not fail, only if the universe collapse or similar. 

Please match here and just end the function if the span is returning `None`.

Reply:
Well I made it a proof because I felt 100% sure it was correct when I read the doc:

> If this returns None, then no span exists for that ID (either it has closed or the ID is invalid).

Otherwise I wouldn't have use the expect @bkchr :sweat: 

I asked on the tracing channel and [Eliza told me the span definitely exists if this method is called so I can unwrap it](https://discord.com/channels/500028886025895936/627649734592561152/801855249001611295).

@dvdplm I'm not sure I understand your thinking here. If you want you can come back on the discord and explain again?

Reply:
In the end I will just do the return because even if there was a bug in tracing and it would fail, we don't want the process to stop because of that.

Reply:
The point is more, when their api is not doing this expect internally, we should not add it :P In the end better to not send some telemetry message, than having a node failing because of this :D

Reply:
Not much thinking at all here, just that I didn't understand the proof. I still don't know why `ctx.span()` returns an `Option`, i.e. under what conditions it's `None`, but I don't need to know. Returning is fine.

Reply:
> I still don't know why ctx.span() returns an Option

It's because `ctx.span(<id>)` returns a span for the id you provided. You can ask any id really, even of a span you disabled or an invalid id (you can make an id from a `u64`). But since `new_span` provides you an id of a `span` that has just been created, there is just no way it would be None. It's like doing an insert on a HashMap and later doing a get on a key you know you inserted. It's just there.